misk/src/main/kotlin/misk/config/MiskConfig.kt

```diff
@@ -113,7 +131,58 @@ object MiskConfig {
     }
   }
 
+  private fun checkConfigStructure(configClass: Class<out Config>) {
+    check(!Secret::class.java.isAssignableFrom(configClass)) { "Top level service config cannot be a Secret<*>" }
+    checkConfigStructureRecursively(configClass, false)
+  }
+
+  @Suppress("UNCHECKED_CAST")
+  private fun checkConfigStructureRecursively(configClass: Class<out Config>, secretFound: Boolean) {
+    for (property in configClass.kotlin.declaredMemberProperties) {
+      if (property.returnType.isSubtypeOf(
+              Secret::class.createType(mutableListOf(KTypeProjection.STAR)))) {
+        check(!secretFound) { "Configs cannot have nested Secret<*>" }
+        val propertyGenericType = configClass.getDeclaredField(property.name).genericType
+
+        val propertyParameterizedType = propertyGenericType as ParameterizedType
+        val propertyType = propertyParameterizedType.actualTypeArguments[0] as Class<*>
+
+        if (Config::class.java.isAssignableFrom(propertyType)) {
+          checkConfigStructureRecursively(propertyType as Class<out Config>, true)
+        }
+
+      } else if (property.returnType.isSubtypeOf(Config::class.createType())) {
+        checkConfigStructureRecursively(
+            property.returnType.typeLiteral().rawType as Class<out Config>, secretFound)
+      }
+    }
+  }
+
   /** @return the list of config file names in the order they should be read */
   private fun embeddedConfigFileNames(appName: String, environment: Environment) =
       listOf("common", environment.name.toLowerCase()).map { "$appName-$it.yaml" }
+
+  private class SecretDeserializer : JsonDeserializer<Secret<*>>(), ContextualDeserializer {
+    private lateinit var thisType: Class<*>
+
+    override fun createContextual(
+      ctxt: DeserializationContext?,
+      property: BeanProperty?
+    ): JsonDeserializer<*> {
+      thisType = property!!.type.bindings.getBoundType(0).rawClass
+      return this
+    }
+
+    override fun deserialize(p: JsonParser?, ctxt: DeserializationContext?): Secret<*> {
+      val reference = ctxt?.readValue(p, String::class.java) as String
+      return DeferredSecret(reference, thisType)
+    }
+  }
+
+  class DeferredSecret<T>(val reference: String, val type: Class<T>) : Secret<T> {
+    var deferredValue: T? = null
+
+    override val value: T
+      get() = deferredValue!!
```

Comment:
We probably want to throw something other than `NullPointerException` if the deferred value is not null, so that it's clear that the caller tried to reference a secret before it was loaded.

Reply:
Also this isn't thread safe.